policy/predicate/status.go

```diff
@@ -16,37 +16,76 @@ package predicate
 
 import (
 	"context"
+	"fmt"
 	"strings"
 
 	"github.com/palantir/policy-bot/policy/common"
 	"github.com/palantir/policy-bot/pull"
 	"github.com/pkg/errors"
 )
 
-type HasSuccessfulStatus []string
+type HasStatus struct {
+	Conclusions allowedConclusions `yaml:"conclusions"`
+	Statuses    []string           `yaml:"statuses"`
+}
 
-var _ Predicate = HasSuccessfulStatus([]string{})
+func NewHasStatus(statuses []string, conclusions []string) *HasStatus {
+	conclusionsSet := make(allowedConclusions, len(conclusions))
+	for _, conclusion := range conclusions {
+		conclusionsSet[conclusion] = unit{}
+	}
+	return &HasStatus{
+		Conclusions: conclusionsSet,
+		Statuses:    statuses,
+	}
+}
 
-func (pred HasSuccessfulStatus) Evaluate(ctx context.Context, prctx pull.Context) (*common.PredicateResult, error) {
-	statuses, err := prctx.LatestStatuses()
+// UnmarshalYAML implements the yaml.Unmarshaler interface for HasStatus.
+// This supports unmarshalling the predicate in two forms:
+//  1. A list of strings, which are the statuses to check for. This is the
+//     deprecated `has_successful_status` format.
+//  2. A full structure with `statuses` and `conclusions` fields as in
+//     `has_status`.
+func (pred *HasStatus) UnmarshalYAML(unmarshal func(interface{}) error) error {
```

Comment:
While this works, it does mean people could use the new `has_status` predicate by providing only a list of strings or use the old `has_successful_status` predicate with a `conclusions` list that doesn't contain `success`.

Instead, I suggest keeping the old `HasSuccessfulStatus` type and replacing the `Evaluate` method with a version that creates a `HasStatus` object and calls `Evaluate` on it. That would keep these as two separate predicates for policy parsing purposes, while avoiding most of the code duplication.

Reply:
Makes sense. Done here: 1793921421e208e47ae226e1d93934152b09bc84.

Since we brought the type back, I added it to the testsuite again. Then it started to look a bit unwieldy with all the cases going on, so I reworked that a bit to be hopefully clearer vs. my previous.